test/Identity.IntegrationTest/Endpoints/IdentityServerTests.cs

```diff
@@ -0,0 +1,298 @@
+                { "scope", "api offline_access" },
+                { "client_id", "web" },
+                { "deviceType", "10" },
+                { "deviceIdentifier", deviceId },
+                { "deviceName", "firefox" },
+                { "grant_type", "password" },
+                { "username", username },
+                { "password", "master_password_hash" },
+            }), context => context.Request.Headers.Add("Auth-Email", CoreHelpers.Base64UrlEncodeString("bad_value")));
+
+            Assert.Equal(StatusCodes.Status400BadRequest, context.Response.StatusCode);
+
+            var body = await AssertHelper.ResponseIsAsync<JsonDocument>(context);
+            var root = body.RootElement;
+
+            var error = AssertHelper.AssertJsonProperty(root, "error", JsonValueKind.String).GetString();
+            Assert.Equal("invalid_grant", error);
+            AssertHelper.AssertJsonProperty(root, "error_description", JsonValueKind.String);
+        }
+
+        [Fact]
+        public async Task TokenEndpoint_GrantTypeRefreshToken_Success()
+        {
+            var deviceId = "5a7b19df-0c9d-46bf-a104-8034b5a17182";
+            var username = "test+tokenrefresh@email.com";
+
+            await _factory.RegisterAsync(new RegisterRequestModel
+            {
+                Email = username,
+                MasterPasswordHash = "master_password_hash",
+            });
+
+            var (_, refreshToken) = await _factory.TokenFromPasswordAsync(username, "master_password_hash", deviceId);
+
+            var context = await _factory.Server.PostAsync("/connect/token", new FormUrlEncodedContent(new Dictionary<string, string>
+            {
+                { "grant_type", "refresh_token" },
+                { "client_id", "web" },
+                { "refresh_token", refreshToken },
+            }));
+
+            await AssertDefaultTokenBodyAsync(context);
+        }
+
+        [Fact]
+        public async Task TokenEndpoint_GrantTypeClientCredentials_Success()
+        {
+            var username = "test+tokenclientcredentials@email.com";
+            var deviceId = "8f14a393-edfe-40ba-8c67-a856cb89c509";
+
+            await _factory.RegisterAsync(new RegisterRequestModel
+            {
+                Email = username,
+                MasterPasswordHash = "master_password_hash",
+            });
+
+            var database = _factory.GetDatabaseContext();
+            var user = await database.Users
+                .FirstAsync(u => u.Email == username);
+
+            var context = await _factory.Server.PostAsync("/connect/token", new FormUrlEncodedContent(new Dictionary<string, string>
+            {
+                { "grant_type", "client_credentials" },
+                { "client_id", $"user.{user.Id}" },
+                { "client_secret", user.ApiKey },
+                { "scope", "api" },
+                { "DeviceIdentifier", deviceId },
+                { "DeviceType", ((int)DeviceType.FirefoxBrowser).ToString() },
```

Comment:
This feels better than hardcoding the device type like you do in the other tests

Reply:
Agreed. I added a little helper to get the number as a string.